tasmota/tasmota_xdrv_driver/xdrv_92_esp32_gnss.ino

```diff
@@ -0,0 +1,876 @@
+    {PSTR("SERIAL_6O2"), SERIAL_6O2},
+    {PSTR("SERIAL_7O2"), SERIAL_7O2},
+    {PSTR("SERIAL_8O2"), SERIAL_8O2}};
+
+/*********************************************************************/
+/*                         HELPER FUNCTIONS                          */
+/*********************************************************************/
+static const char *SerialConfigToString(uint32_t config)
+{
+  for (const auto &mapping : kSerialConfigs)
+  {
+    if (mapping.config == config)
+    {
+      return mapping.name;
+    }
+  }
+  return "UNKNOWN";
+}
+
+static uint32_t StringToSerialConfig(const char *name)
+{
+  for (const auto &mapping : kSerialConfigs)
+  {
+    if (strcasecmp(mapping.name, name) == 0)
+    {
+      return mapping.config;
+    }
+  }
+  return SERIAL_8N1;
+}
+
+/*********************************************************************/
+/*                   LOAD/SAVE SETTINGS FUNCTIONS                    */
+/*********************************************************************/
+void GNSSSettingsSave(void)
+{
+  uint32_t crc32 = GetCfgCrc32((uint8_t *)&GNSSSettings + 4, sizeof(GNSSSettings) - 4);
+  if (crc32 != GNSSSettings.crc32)
+  {
+    GNSSSettings.crc32 = crc32;
+
+    char filename[20];
+    snprintf_P(filename, sizeof(filename), PSTR(TASM_FILE_DRIVER), XDRV_92);
+
+    AddLog(LOG_LEVEL_DEBUG, PSTR("GNSS: Saving settings to file %s"), filename);
+
+#ifdef USE_UFILESYS
+    if (!TfsSaveFile(filename, (const uint8_t *)&GNSSSettings, sizeof(GNSSSettings)))
+    {
+      AddLog(LOG_LEVEL_INFO, D_ERROR_FILE_NOT_FOUND);
+    }
+#else
+    AddLog(LOG_LEVEL_INFO, D_ERROR_FILESYSTEM_NOT_READY);
+#endif // USE_UFILESYS
+  }
+}
+
+void GNSSSettingsLoad(void)
+{
+  char filename[20];
+  snprintf_P(filename, sizeof(filename), PSTR(TASM_FILE_DRIVER), XDRV_92);
+
+#ifdef USE_UFILESYS
+  if (TfsLoadFile(filename, (uint8_t *)&GNSSSettings, sizeof(GNSSSettings)))
+  {
+    uint32_t crc32 = GetCfgCrc32((uint8_t *)&GNSSSettings + 4, sizeof(GNSSSettings) - 4);
+    if (crc32 != GNSSSettings.crc32)
+    {
+      memset(&GNSSSettings, 0, sizeof(GNSSSettings));
+      GNSSSettings.version = 1;
+      GNSSSettings.baudrate = 115200;
+      GNSSSettings.serial_config = SERIAL_8N1;
+    }
+  }
+  else
+  {
+    memset(&GNSSSettings, 0, sizeof(GNSSSettings));
+    GNSSSettings.version = 1;
+    GNSSSettings.baudrate = 115200;
+    GNSSSettings.serial_config = SERIAL_8N1;
+  }
+#endif // USE_UFILESYS
+
+  GNSSData.baudrate = GNSSSettings.baudrate;
+  GNSSData.serial_config = GNSSSettings.serial_config;
+}
+
+/*********************************************************************/
+/*                          CORE FUNCTIONS                           */
+/*********************************************************************/
+void GNSSInit(void)
+{
+  if (!GNSSSettings.baudrate)
+  {
+    GNSSSettingsLoad();
+  }
+
+  if (gnssSerial)
+  {
+    delete gnssSerial;
+    gnssSerial = nullptr;
+  }
+
+  gnssSerial = new HardwareSerial(1);
```

Comment:
Please use TasmotaSerial class instead. It abstract all the details of serial management. It may as well make it working on ESP8266.